utils/build_rpms_from_scratch.sh

```diff
@@ -6,52 +6,126 @@ set -uex
 setup_local_repo() {
   local rpms_dst=$1
 
+  if [[ "$OS_ID" == "opensuse-leap" ]]; then
+    repo_file=/etc/zypp/repos.d/local-rpms.repo
+  else
+    repo_file=/etc/yum.repos.d/local-rpms.repo
+  fi
+
   echo "[local-rpms]
 name=Local RPM Repository
 baseurl=file://$rpms_dst
 enabled=1
-gpgcheck=0" > /etc/yum.repos.d/local-rpms.repo
+gpgcheck=0" > "$repo_file"
 }
 
 # Clean up the temporary repository configuration
 cleanup_local_repo() {
-  rm -f /etc/yum.repos.d/local-rpms.repo
+  if [[ "$OS_ID" == "opensuse-leap" ]]; then
+    rm -f /etc/zypp/repos.d/local-rpms.repo
+  else
+    rm -f /etc/yum.repos.d/local-rpms.repo
+  fi
 }
 
-dnf_builddep() {
-  local rpms_dst=$1
-  local spec_file="$2"
-  # dnf --nogpgcheck --repofrompath rpms,"$rpms_dst" --refresh builddep -y "$spec_file"
-  dnf --refresh builddep -y "$spec_file"
+install_dependencies() {
+  local spec_file="$1"
+
+  if [[ "$OS_ID" == "opensuse-leap" ]]; then
+   dependencies=$(rpmspec -q --buildrequires "$spec_file" | grep -v "^rpmlib" | grep -v "^/" | tr '\n' ' ')
+      if [ -n "$dependencies" ]; then
+          # Install dependencies using zypper
+          echo "Installing build dependencies: $dependencies"
+          zypper install --allow-vendor-change -y $dependencies
+      else
+          echo "No build dependencies found in the spec file."
+      fi
+  else
+    dnf --refresh builddep -y "$spec_file"
+  fi
+}
+
+prepare_env_leap() {
+  suse_version=$OS_VERSION # default to '15.4'
+
+  # Update and install required packages using zypper
+  zypper refresh
+
+  # Install needed packages
+  # git so can clone the repo
+  # rpm-build, rpmdevtools, createrepo for building the rpms and setting up the rpm repo
+  # make is used for managing the rpmbuild process. This should probably be a dependency of the rpm.spec files, but not
+  # python3-Sphinx is needed by dpdk. The file /usr/bin/sphinx-build is listed as a dependency in the spec file,
+  #   however, zypper isn't able to resolve a filename to a package and the scripting to handle that case is pretty
+  #   messy, especially because "zypper what-provides" actually returns three options for sphinx-build. So, forgoing
+  #   automation here and simply hardcoding the dependency.
+  zypper install -y \
+      git \
+      rpm-build \
+      rpmdevtools \
+      createrepo_c \
+      make \
+      python3-Sphinx
+
+
+  # Install build tools and dependencies
+  zypper install -y -t pattern devel_basis
+
+  # Enable additional repositories if needed
+  if ! zypper lr | grep -q "Main Repository"; then
+      zypper ar -f http://download.opensuse.org/distribution/leap/$suse_version/repo/oss/ "Main Repository"
+  else
+      echo "Main Repository already exists, skipping addition."
+  fi
+  if ! zypper lr | grep -q "Main Update Repository"; then
+      zypper ar -f http://download.opensuse.org/update/leap/$suse_version/oss/ "Main Update Repository"
+  else
+      echo "Main Update Repository already exists, skipping addition."
+  fi
```

Comment:
Just a bit of DRY here:
```suggestion
  for repo in "Main Repository" "Main Update Repository"; do
    if ! zypper lr | grep -q "$repo"; then
        zypper ar -f http://download.opensuse.org/distribution/leap/$suse_version/repo/oss/ "$repo"
    else
        echo "Main Repository already exists, skipping addition."
    fi
  done
```

Reply:
The URLs need to be different as well, right? I don't know bash as well, is there a way to iterate over a key/value pair? I know, this is my Python mind working. :) 
"Main Repository": http://download.opensuse.org/distribution/leap/$suse_version/repo/oss/
"Main Update Repository": http://download.opensuse.org/update/leap/$suse_version/oss/

Reply:
> The URLs need to be different as well, right?

Ahhh.  I missed that.

> I don't know bash as well, is there a way to iterate over a key/value pair?

Yes, using associative arrays.  Something like:

```bash
declare -A repos
repos["Main Repository"]="distribution/leap/$suse_version/repo/oss
repos["Main Update Repository"]="update/leap/$suse_version/oss"

for repo in "${!repos[@]}"; do
  if ! zypper lr | grep -q "$repo"; then
    zypper ar -f http://download.opensuse.org/"${repos[$repo]}" "$repo"
  else
    echo "$repo already exists, skipping addition."
  fi
done
```